test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc

```diff
@@ -2014,4 +2014,61 @@ TEST_P(ExtProcIntegrationTest, SendHeaderAndTrailerInBufferedMode) {
   verifyDownstreamResponse(*response, 200);
 }
 
+// Test the filter with header allow list configured and verify only the allowed headers
+// in request or response headers and trailers are sent to the ext_proc server.
+TEST_P(ExtProcIntegrationTest, GetAndSetHeadersAndTrailersWithHeaderScrubbing) {
+  auto* forward_rules = proto_config_.mutable_forward_rules();
+  auto* list = forward_rules->mutable_allowed_headers();
+  list->add_patterns()->set_exact(":method");
+  list->add_patterns()->set_exact(":authority");
+  list->add_patterns()->set_exact(":status");
+  list->add_patterns()->set_exact("x-test-trailers");
+  proto_config_.mutable_processing_mode()->set_request_header_mode(ProcessingMode::SEND);
+  proto_config_.mutable_processing_mode()->set_request_trailer_mode(ProcessingMode::SEND);
+  proto_config_.mutable_processing_mode()->set_response_header_mode(ProcessingMode::SEND);
+  proto_config_.mutable_processing_mode()->set_response_trailer_mode(ProcessingMode::SEND);
+
+  initializeConfig();
+  HttpIntegrationTest::initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  Http::TestRequestHeaderMapImpl headers;
+  HttpTestUtility::addDefaultHeaders(headers);
+  auto encoder_decoder = codec_client_->startRequest(headers);
+  request_encoder_ = &encoder_decoder.first;
+  IntegrationStreamDecoderPtr response = std::move(encoder_decoder.second);
+  Http::TestRequestTrailerMapImpl request_trailers{{"x-trailer-foo", "yes"}};
+  codec_client_->sendTrailers(*request_encoder_, request_trailers);
+
+  processRequestHeadersMessage(
+      *grpc_upstreams_[0], true, [](const HttpHeaders& headers, HeadersResponse&) {
+        Http::TestRequestHeaderMapImpl expected_request_headers{{":method", "GET"},
+                                                                {":authority", "host"}};
+        // Verify only allowed request headers is received by ext_proc server.
+        EXPECT_THAT(headers.headers(), HeaderProtosEqual(expected_request_headers));
+        return true;
+      });
+  processRequestTrailersMessage(*grpc_upstreams_[0], false,
```

Comment:
if `handleUpstreamRequestWithTrailer` is not called, it is expected that no trailer is there even without allowList rule, right?
Probably a better test will be: encode the other trailer besides `x-test-trailers` and verify that is not there.

Reply:
codec_client->sendTrailers("x-trailer-foo) will send downstream request trailers to ext_proc filter, and be forwarded to ext_proc server. However, since "x-trailer-foo" is not in the allow header list, so the trailer forwarded to ext_proc server doesn't contain any header.  This is tested and verified the allow header feature for the request trailer.

With handleUpstreamRequestWithTrailer(), the upstream will send back and response with trailer which contains a fixed header: x-test-trailers. Since it is in allow list, it will be forwarded to the ext_proc server. And ext_proc server verified this. This tested the allow header feature for the response trailer.

Combine this two, it tested the scenario you mentioned. Also, with the header allow list for request / response headers, I feel the test is comprehensive. Otherwise, we will have to modify handleUpstreamRequestWithTrailer() a little bit to add more headers there, but is not necessary.

Reply:
ah, I missed single line `sendTrailers("x-trailer-foo)` above